File: examples/04-kafka/python/signal_bridge.py
from temporalio.client import Client

# `OrderWorkflow` is the workflow class with @workflow.run run(order_id) and a
# @workflow.signal order_event(payload). `record` is a kafka-python ConsumerRecord.


# start_workflow(..., start_signal=...) is the Python equivalent of Java's
# signalWithStart: it starts the Workflow on the first event for a key, and the
# attached signal is delivered to that same execution. A re-running id reuse
# policy means later events signal the existing run rather than erroring.
async def on_kafka_record(client: Client, record) -> None:
    order_id = record.key.decode()
    await client.start_workflow(
        OrderWorkflow.run,
        order_id,
        id=f"order-{order_id}",
        task_queue="orders",
        start_signal="order_event",
        start_signal_args=[record.value.decode()],
    )

File: examples/04-kafka/go/signal_bridge.go
package kafka

import (
	"context"

	"github.com/segmentio/kafka-go"
	"go.temporal.io/sdk/client"
)

// SignalWithStartWorkflow is the Go equivalent of Java's signalWithStart: it
// starts the Workflow on the first event for a key and signals the existing
// execution for every event after that. A fixed WorkflowID derived from the
// record key means the same key always maps to the same execution.
func OnKafkaRecord(ctx context.Context, c client.Client, record kafka.Message) error {
	orderID := string(record.Key)
	_, err := c.SignalWithStartWorkflow(
		ctx,
		"order-"+orderID,     // WorkflowID
		"orderEvent",         // signal name
		string(record.Value), // signal arg
		client.StartWorkflowOptions{ID: "order-" + orderID, TaskQueue: "orders"},
		"OrderWorkflow", // workflow type
		orderID,         // workflow arg
	)
	return err
}
